presto-main/src/main/java/com/facebook/presto/execution/scheduler/AdaptivePhasedExecutionPolicy.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler;
+
+import com.facebook.presto.Session;
+
+import java.util.Collection;
+
+import static com.facebook.presto.SystemSessionProperties.getMaxStageCountForEagerScheduling;
+
+public class AdaptivePhasedExecutionPolicy
+        implements ExecutionPolicy
+{
+    @Override
+    public ExecutionSchedule createExecutionSchedule(Session session, Collection<StageExecutionAndScheduler> stages)
+    {
+        if (stages.size() > getMaxStageCountForEagerScheduling(session)) {
+            return new PhasedExecutionSchedule(stages);
+        }
+        else {
+            return new AllAtOnceExecutionSchedule(stages);
+        }
+    }
+}
```

Comment:
LGTM % a noob question.
Can the same be achieved in the optimizer by setting the [query.execution-policy](https://github.com/prestodb/presto/blob/2eeaf68ed36768adb504617b0e41a30253a94b4a/presto-main/src/main/java/com/facebook/presto/execution/QueryManagerConfig.java#L479-L484) when the plan is being prepared

Reply:
@ajaygeorge do you have suggestions on how that code change might look like?

Reply:
I have no idea regarding the actual implementation. It was more of a noob question. :) 
Thinking in high level terms this felt like an "optimization" step which can be done while preparing the plan vs doing it during scheduling/execution. 
Just grepping through the code there seems to be a bunch of [runtimePlanOptimizers](https://github.com/prestodb/presto/blob/b496843cf52a132780776dd9db00980d4879abda/presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java#L608) which seems like a place where we could plug this in.

Reply:
The intent here is different: we presume that we have fully optimized the plan, but identify a mechanism to schedule the stages such that we reduce the scheduled stages to only what is needed to prevent deadlock and make progress in the query.  So to me this doesn't seem like an optimization, it's a different scheduling mechanism, and it seems sensible to expose this as a tunable property of the cluster or query.

Reply:
The intent here is different: we presume that we have fully optimized the plan, but identify a mechanism to schedule the stages such that we reduce the scheduled stages to only what is needed to prevent deadlock and make progress in the query.  So to me this doesn't seem like an optimization, it's a different scheduling mechanism, and it seems sensible to expose this as a tunable property of the cluster or query.

Reply:
yeah this isn't really something the optimizer sets.  I think this approach is good.